### DISCUSSION TITLE
System timer alarms of RP2040, anyone?

### DISCUSSION THREAD

**Discussion by cve2022** [micropython/micropython#10713] [RP2040 / Pico]
I am thinking of using the alarms of this timer to generate a periodic interrupt. My idea is using the alarm ISR (there are up to four alarms according to the data sheet) as a poor man's RTOS, to execute some short tasks as fast as possible (after the alarm event) and with as low a jitter as possible.

Soo, the first question is: has anybody done that already? I would hate to reinvent the wheel.

**Comment by freemansoft** [comment]
How is the alarm different from a timer interrupt?  A quick look at https://raspberrypi.github.io/pico-sdk-doxygen/group__hardware__timer.html makes me wonder if they are the same.  Micropython supports hardware timer callbacks.

**Reply by cve2022** [reply]
> How is the alarm different from a timer interrupt?

My understanding is they are exactly the same. I was just wanting to know if somebody had already had some MP code I could take as is or adapt it to my needs.

I suspect the Timer class uses at least one of the alarms and somehow it serializes the interrupts when several Timer objects are used, instead of using one alarms for each instance of the Timer class, as I was able to create much more than four Timer objects without any error.

I played with the Timer class and it works fine (as long as you do not abuse it :smile:, but I only tested it for simple code (and one abusive test code). I do not know how it would perform in a more complex application where more tasks are running and possibly competing for the timer resources. So I thought I could have one alarm dedicated for the most realtime sensitive part of my application, to hopefully reduce its latency and jitter.

**Reply by freemansoft** [reply]
Hmm.  I think that Micropython only implements a software timer on the RP2040 and the ESP8266.  So you are essentially on a virtual timer.  The ESP32C3 has hardware timer support https://docs.micropython.org/en/latest/esp32/quickref.html?highlight=timer%20esp32#timers .   At least that was true a couple of months ago when I worked on It and at my notes in main.py https://github.com/freemansoft/ESP8266-MicroPython  

> Timers
> RP2040’s system timer peripheral provides a global microsecond timebase and generates interrupts for it. The software timer is available currently, and there are unlimited number of them (memory permitting). There is no need to specify the timer id (id=-1 is supported at the moment) as it will default to this.

**Reply by cve2022** [reply]
> Hmm. I think that Micropython only implements a software timer on the RP2040 and the ESP8266. So you are essentially on a virtual timer. 

Not sure what you mean by "software timer" and "virtual timer", but I would say it certainly uses the system timer.

I have done a small experiment:

```
from micropython import const
from machine import Pin, mem32, Timer

TIMER_BASE = const(0x40054000)
TIMEHW = const (TIMER_BASE + 0x00)
TIMELW = const (TIMER_BASE + 0x04)
TIMEHR = const (TIMER_BASE + 0x08)
TIMELR = const (TIMER_BASE + 0x0C)
ALARM0 = const (TIMER_BASE + 0x10)
ALARM1 = const (TIMER_BASE + 0x14)
ALARM2 = const (TIMER_BASE + 0x18)
ALARM3 = const (TIMER_BASE + 0x1C)
ARMED =  const (TIMER_BASE + 0x20)
TIMERAWH = const (TIMER_BASE + 0x24)
TIMERAWL = const (TIMER_BASE + 0x28)
DBGPAUSE = const (TIMER_BASE + 0x2C)
PAUSE = const (TIMER_BASE + 0x30)
INTR = const (TIMER_BASE + 0x34)
INTE = const (TIMER_BASE + 0x38)
INTF = const (TIMER_BASE + 0x3C)
INTS = const (TIMER_BASE + 0x40)

TMR_LIST = const ( (
    'TIMELR', 'ALARM0', 'ALARM1', 'ALARM2', 'ALARM3', 'ARMED',
    'INTR', 'INTE', 'INTF', 'INTS',
) )

led = Pin(25, Pin.OUT)
ld0 = Pin(0, Pin.OUT)
ld1 = Pin(1, Pin.OUT)

def cb (t=None):
    led.toggle()
def cb0 (t=None):
    ld0.toggle()
def cb1 (t=None):
    ld1.toggle()
t = Timer()
t0 = Timer()
t1 = Timer()
t.init(mode = Timer.PERIODIC, freq = 10, callback = cb)
t0.init(mode = Timer.PERIODIC, freq = 7, callback = cb0)
t1.init(mode = Timer.PERIODIC, freq = 13, callback = cb1)

for i in TMR_LIST:
    v = eval(i)
    print (i + '-> ' + hex(v) + ': ' + hex(mem32[v]))
```
Result:
```
TIMELR-> 0x4005400c: 0x318c8102
ALARM0-> 0x40054010: 0x0
ALARM1-> 0x40054014: 0x0
ALARM2-> 0x40054018: 0x0
ALARM3-> 0x4005401c: 0x318da5c0
ARMED-> 0x40054020: 0x8
INTR-> 0x40054034: 0x0
INTE-> 0x40054038: 0x8
INTF-> 0x4005403c: 0x0
INTS-> 0x40054040: 0x0
```

My preliminary conclusions are:
1- MP uses the ALARM3 interrupt to do its timing stuf, including the Timer objects. The result is similar even if I do not create the three Timer objects.
2- I believe the timer subsystem keeps a list of timing objects, including the Timers and sets the ALARM3 register for the next in list that will timeout.

I may use one of the other three ALARMs to do what I want, at least with the current version of MP.

**Reply by cve2022** [reply]
So... I think I figured it out.

My ALARM interrupt callback will need to update the ALARMx register for next timeout and rearm the interrupt  (possibly just cleaning the interrupt flag) and then calling a callback that will do the processing.

So, the next question is how can I set this callback in python? Is there already a function that I may call that sets the interrupt controller (NVIC)?

**Comment by freemansoft** [comment]
I meant that in the MicroPython for RP2040.  They call it a software timer rather than hardware.  It was in the quote of the earlier message which is from the micropython rp2040 docs

**Reply by cve2022** [reply]
> I meant that in the MicroPython for RP2040. They call it a software timer rather than hardware. It was in the quote of the earlier message which is from the micropython rp2040 docs

Oh, I see.

**Comment by redhead-p** [comment]
This comes with a health warning as I haven't tried it.   The following PIO state machine example shows how to generate a 1Hz timer and link it to a ISR.  It should be possible to adjust this to any reasonable frequency.

[https://github.com/raspberrypi/pico-micropython-examples/blob/master/pio/pio_1hz.py](url)

There are other examples for 2000Hz.  The examples run the ISRs in a soft (interruptible) context, but you could change this to hard (uninterruptible) by adding hard = True if desired. E.g
`
sm.irq(lambda p: print(time.ticks_ms()), hard = True)
`

This would avoid any complications associated with using the timer peripheral or setting the interrupt up yourself with the  NVIC

**Reply by cve2022** [reply]
> This comes with a health warning as I haven't tried it. The following PIO state machine example shows how to generate a 1Hz timer and link it to a ISR. It should be possible to adjust this to any reasonable frequency.
> 

Thanks for pointing this up, I will take a look at it.

FWIW the link above gets a 404 page.

[https://github.com/raspberrypi/pico-micropython-examples/tree/master/pio](https://github.com/raspberrypi/pico-micropython-examples/tree/master/pio)
